R/summarise_scores.R

```diff
@@ -0,0 +1,112 @@
+#' Summarise scores
+#'
+#' @param scores table of scores, e.g. as created with [score_forecasts()]
+#' @param report_date Date at which the scoring takes place
+#' @param restrict_weeks Integer number of continuous weeks continuous weeks
+#' leading up to the `report_date` the forecasts need to include to be
+#' considered in the scoring.
+#'
+#' @importFrom dplyr group_by mutate ungroup filter select bind_rows count summarise left_join select across n_distinct full_join distinct starts_with
+#' @importFrom tidyr complete replace_na
+#' @importFrom lubridate weeks
+#' @importFrom scoringutils pairwise_comparison
+#'
+#' @export
+summarise_scores <- function(scores, report_date, restrict_weeks = 0L) {
+
+  last_forecast_date <- report_date - 7
+
+  locations <- scores %>%
+    select(location, location_name) %>%
+    distinct()
+
+  ## extract data to be scored and set number of locations to one as default (see next command)
+  score_data <- scores %>%
+    filter(forecast_date <= last_forecast_date,
+           target_end_date <= report_date)
+
+  ## duplicate country data as overall data
+  score_df <- score_data %>%
+    mutate(location = "Overall") %>%
+    bind_rows(score_data)
+
+  num_loc <- score_df %>%
+    group_by(model, location, target_variable, horizon) %>%
+    summarise(n_loc = n_distinct(location_name), .groups = "drop")
+
+  ## for overall, if more than 1 location exists, filter to have at least half
+  ## of them
+  score_df <- score_df %>%
+    group_by(model, target_variable, location, horizon) %>%
+    mutate(n = n_distinct(location_name)) %>%
+    ungroup() %>%
+    mutate(nall = n_distinct(location_name)) %>%
+    filter(location != "Overall" | n >= nall / 2) %>%
+    select(-n, -nall)
+
+  ## continuous weeks of submission
+  if (restrict_weeks > 0) {
+    cont_weeks <- score_df %>%
+      group_by(model, location, target_variable, target_end_date, horizon, forecast_date) %>%
+      summarise(present = 1, .groups = "drop") %>%
+      complete(model, location, target_variable, target_end_date, horizon) %>%
```

Comment:
We used to create the complete set of possibilities with:

```r
complete(model, location, target_variable, forecast_date, horizon)
```

Upon cursory look, it seems equivalent to the current approach but it's not. `complete()` doesn't behave like `group_by()`.

I *believe* (but not 100% sure. I'm having a hard time wrapping my head around this) that `forecast_date` is the right option since we want to make sure that a forecast was posted each week, not that each week was forecasted.